base/linalg/diagonal.jl

```diff
@@ -36,15 +36,22 @@ fill!(D::Diagonal, x) = (fill!(D.diag, x); D)
 
 full(D::Diagonal) = diagm(D.diag)
 
-getindex(D::Diagonal, i::Int, j::Int) = (checkbounds(D, i, j); unsafe_getindex(D, i, j))
-unsafe_getindex{T}(D::Diagonal{T}, i::Int, j::Int) = i == j ? unsafe_getindex(D.diag, i) : zero(T)
-unsafe_getindex{T}(D::Diagonal{Matrix{T}}, i::Int, j::Int) = i == j ? D.diag[i] : zeros(T, size(D.diag[i
-], 1), size(D.diag[j], 2))
+@inline function getindex(D::Diagonal, i::Int, j::Int)
+    @boundscheck checkbounds(D, i, j)
+    if i == j
+        @inbounds r = D.diag[i]
+    else
+        r = diagzero(D, i, j)
```

Comment:
Off-topic for this PR, but wow is this a specialized method that makes an awful lot of assumptions about what you're trying to do...